src/main/kotlin/org/rust/lang/core/types/RustTypeParameterType.kt

```diff
@@ -16,6 +16,8 @@ class RustTypeParameterType(val parameter: RustTypeParamElement) : RustTypeBase(
     override fun getNonStaticMethodsIn(project: Project): Sequence<RustFnElement> =
         getTraitsImplementedIn(project).flatMap { it.traitMethodMemberList.asSequence() }
 
+    override fun substitute(map: Map<RustTypeParameterType, RustType>): RustType = map[this] ?: this
+
```

Comment:
If we are mapping over `RustTypeParameterType` then we should provide specialized `hashCode`,

or I don't know something about Kotlin/Java in this area